src/any_agent/tracing/instrumentation/langchain.py

```diff
@@ -0,0 +1,190 @@
+        if token_usage := getattr(response, "llm_output", {}).get("token_usage"):
+            span.set_attributes(
+                {
+                    "gen_ai.usage.input_tokens": token_usage.prompt_tokens,
+                    "gen_ai.usage.output_tokens": token_usage.completion_tokens,
+                }
+            )
+        span.set_status(StatusCode.OK)
+        span.end()
+
+        del self._current_spans["model"][str(run_id)]
+
+    def on_tool_end(
+        self,
+        output: Any,
+        *,
+        run_id: UUID,
+        parent_run_id: UUID | None = None,
+        **kwargs: Any,
+    ) -> Any:
+        span = self._current_spans["tool"][str(run_id)]
+
+        if output.content:
+            span.set_attributes(
+                {
+                    "genai.output": json.dumps(
+                        output.content,
+                        default=str,
+                        ensure_ascii=False,
+                    ),
+                    "genai.output.type": "json",
+                    "gen_ai.tool.call.id": output.tool_call_id,
+                }
+            )
+
+        span.set_status(StatusCode.OK)
+        span.end()
+
+        del self._current_spans["tool"][str(run_id)]
+
+
+class _LangChainInstrumentor:
+    def instrument(self, tracer: Tracer) -> None:
+        tracing_callback = _LangChainTracingCallback(tracer)
+
+        self._config_set = False
+
+        async def wrap_ainvoke(  # type: ignore[no-untyped-def]
+            wrapped: Callable[..., None],
+            instance: Any,
+            args: Any,
+            kwargs: Any,
+        ):
+            if not self._config_set:
+                if "config" in kwargs:
+                    if callbacks := kwargs["config"].get("callbacks"):
+                        if isinstance(callbacks, list):
+                            kwargs["config"]["callbacks"].append(tracing_callback)
+                        else:
+                            original_callback = kwargs["config"]["callbacks"]
+                            kwargs["config"]["callbacks"] = [
+                                original_callback,
+                                tracing_callback,
+                            ]
+                    else:
+                        kwargs["config"]["callbacks"] = [tracing_callback]
+                else:
+                    kwargs["config"] = RunnableConfig(callbacks=[tracing_callback])
+                self._config_set = True
+
+            return await wrapped(*args, **kwargs)  # type: ignore[func-returns-value]
+
+        import langgraph
+
+        self._original_ainvoke = langgraph.pregel.Pregel.ainvoke  # type: ignore[attr-defined]
+        wrap_function_wrapper(  # type: ignore[no-untyped-call]
+            "langgraph.pregel", "Pregel.ainvoke", wrapper=wrap_ainvoke
+        )
+
+    def uninstrument(self) -> None:
+        import langgraph
+
+        langgraph.pregel.Pregel.ainvoke = self._original_ainvoke  # type: ignore[attr-defined]
```

Comment:
Should there be any sort of handling if someone calls uninstrument before they call instrument? Seems like doing that will make this crash.

This comment applies to a few other frameworks too

Reply:
> Should there be any sort of handling if someone calls uninstrument before they call instrument?

How could this happen? The instrumentation happens during agent creation.
Someone would need to manually import the instrumentor (which is marked as `_private`), instantiate it and call `uninstrument`?

Reply:
Yeah I'm referring to that second edge case. It's marked private but that hasn't always stopped their usage. For those using _private fields it's sort of a "use at your own risk" scenario so I'm ok with keeping it as-is, but wanted to draw attention that this is technically a corner case.

Reply:
will update the code just for extra safety

Reply:
updated code and added tests